File: src/main/java/org/c3lang/intellij/psi/impl/C3StructDeclarationMixinImpl.java
package org.c3lang.intellij.psi.impl;

import com.intellij.lang.ASTNode;
import com.intellij.psi.stubs.IStubElementType;
import com.intellij.psi.tree.IElementType;
import org.c3lang.intellij.psi.*;
import org.c3lang.intellij.stubs.C3StructDeclarationStub;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import java.util.Collections;
import java.util.List;

public abstract class C3StructDeclarationMixinImpl extends C3StubBasedPsiElementBase<C3StructDeclarationStub> implements C3StructDeclaration
{
	public C3StructDeclarationMixinImpl(@NotNull ASTNode node)
	{
		super(node);
	}

	public C3StructDeclarationMixinImpl(@NotNull C3StructDeclarationStub stub, @NotNull IStubElementType<?, ?> nodeType)
	{
		super(stub, nodeType);
	}

	public C3StructDeclarationMixinImpl(@NotNull C3StructDeclarationStub stub, @Nullable IElementType nodeType, @Nullable ASTNode node)
	{
		super(stub, nodeType, node);
	}

	@Override
	public @NotNull List<StructField> getFields()
	{
		C3StructDeclarationStub s = getGreenStub();
		if (s != null) return s.getFields();
		C3StructBody structBody = getStructBody();
		if (structBody == null) return Collections.emptyList();
		return StructField.collectFields(structBody, null);
	}

	@Override
	public @NotNull FullyQualifiedName getDeclaredIn()
	{
		C3TypeName typeName = getTypeName();
		return FullyQualifiedName.from(typeName, typeName.getModuleName());
	}
}

File: src/main/java/org/c3lang/intellij/psi/impl/C3StructDeclarationMixinImpl.kt
package org.c3lang.intellij.psi.impl

import com.intellij.lang.ASTNode
import com.intellij.psi.stubs.IStubElementType
import com.intellij.psi.tree.IElementType
import org.c3lang.intellij.psi.C3StructDeclaration
import org.c3lang.intellij.psi.FullyQualifiedName
import org.c3lang.intellij.psi.StructField
import org.c3lang.intellij.psi.StructField.Companion.collectFields
import org.c3lang.intellij.stubs.C3StructDeclarationStub

abstract class C3StructDeclarationMixinImpl : C3StubBasedPsiElementBase<C3StructDeclarationStub>, C3StructDeclaration {

    constructor(node: ASTNode) : super(node)

    constructor(stub: C3StructDeclarationStub, nodeType: IStubElementType<*, *>) : super(stub, nodeType)

    constructor(stub: C3StructDeclarationStub, nodeType: IElementType?, node: ASTNode?) : super(stub, nodeType, node)

    override val fields: List<StructField>
        get() = greenStub?.fields ?: structBody?.collectFields(null) ?: emptyList()

    override val declaredIn: FullyQualifiedName?
        get() = this.typeName?.let {
            FullyQualifiedName.from(it)
        }
}